File: algorithms/020.valid-parentheses.cpp
class Solution {
public:
    bool is_match(char o, char c){
        if ((o == '(' && c == ')') || (o == '{' && c == '}') || (o == '[' && c == ']'))
            return true;
        return false;
    }

    bool isValid(string s) {
        stack<char> st;
        if (s.size() == 0)
            return true;
            
        for (int i=0; i < s.size(); i++){
            if (s[i] == '(' || s[i] == '{' || s[i] == '[')
                st.push(s[i]);
            else if (st.empty() || !is_match(st.top(),s[i]))
                return false;
            else
                st.pop();
        }

        // if (st.empty())
        //     return true;
        // else
        //     return false;
		return st.empty();
    }
};

File: algorithms/020.valid-parentheses.js
/**
 * @param {string} s
 * @return {boolean}
 */
var isValid = function(s) {
	const stack = [];
	const mapping = {
		')' : '(',
		'}' : '{',
		']' : '['
	};
	for (let i = 0; i < s.length; i++) {
		const char = s[i];
		if (char === '(' || char === '{' || char === '[') {
			stack.push(char);
		} else {
			if (stack.length === 0 || stack[stack.length - 1] != mapping[char])
				return false;
			stack.pop();
		}
	}
	return stack.length === 0;
};
